syncher/syncserver.go

```diff
@@ -25,51 +26,92 @@ import (
 	"google.golang.org/grpc"
 )
 
+const PolicySyncRetryTime = 500 * time.Millisecond
+
 type syncClient struct {
 	target   string
 	dialOpts []grpc.DialOption
 }
 
 type SyncClient interface {
 
-	// Sync connects to the Policy Sync API server and processes updates from it.  It modifies the provided store with
-	// the updates.  Sync blocks until the connection to the API server is terminated.
-	Sync(ctx context.Context, store *policystore.PolicyStore)
+	// Sync connects to the Policy Sync API server and processes updates from it.  It sends PolicyStores over the
+	// channel when they are ready for enforcement.  Each time we disconnect and resync with the Policy Sync API a new
+	// PolicyStore is created.
+	Sync(ctx context.Context, stores chan<- *policystore.PolicyStore)
 }
 
 // NewClient creates a new syncClient.
 func NewClient(target string, opts []grpc.DialOption) SyncClient {
 	return &syncClient{target: target, dialOpts: opts}
 }
 
-func (s *syncClient) Sync(cxt context.Context, store *policystore.PolicyStore) {
+func (s *syncClient) Sync(cxt context.Context, stores chan<- *policystore.PolicyStore) {
+	for {
+		select {
+		case <-cxt.Done():
+			return
+		default:
+			store := policystore.NewPolicyStore()
+			inSync := make(chan struct{})
+			done := make(chan struct{})
+			go s.syncStore(cxt, store, inSync, done)
+
+			// Block until we receive InSync message, or cancelled.
+			select {
+			case <-inSync:
+				stores <- store
+			// Also catch the case where syncStore ends before it gets an InSync message.
+			case <-done:
+				// pass
+			case <-cxt.Done():
+				return
+			}
+
+			// Block until syncStore() ends (e.g. disconnected), or cancelled.
+			select {
+			case <-done:
+				// pass
```

Comment:
not sure but if the above `select` has the `case <-done:` then is it ok that we are again checking for `case <-done` here?

Reply:
this is fine, because we close the `done` channel, rather than writing to it. Reading from a closed channel does not block and returns the zero value for the channel data type.